### PR TITLE
Download artifacts functionality added

### PR DESCRIPTION
**Pull request by shiv-am0**

### CODE REVIEW

File: docker/Dockerfile.linux.arm64 [hunk 1]
```diff
@@ -1,7 +1,7 @@
 FROM arm64v8/alpine:3.15 as alpine
 RUN apk add -U --no-cache ca-certificates
 
-FROM arm64v8/alpine:3.15
+FROM scratch
```
Comment (vistaarjuneja) [inline docker/Dockerfile.linux.arm64:4]: Have some common comments between the s3 PR and this one - please refer to that PR once

File: plugin.go [hunk 2]
```diff
@@ -78,6 +83,73 @@ func (p *Plugin) Exec(client *storage.Client) error {
...
+		ctx := context.Background()
+		query := &storage.Query{Prefix: p.Config.Source}
+
+		// List the objects in the specified GCS bucket path
+		it := p.bucket.Objects(ctx, query)
+
+		g := errgroup.Group{}
+
+		for {
+			objAttrs, err := it.Next()
```
Comment (vistaarjuneja) [inline plugin.go:104]: Move the downloading to a different function

File: plugin.go [hunk 3]
```diff
@@ -282,3 +354,17 @@ func (p *Plugin) walkFiles() ([]string, error) {
 
 	return items, err
 }
+
+func extractDir(destination string) string {
```
Comment (vistaarjuneja) [inline plugin.go:358]: this is just a line function, you can remove it and use `filepath.Dir(destination)` directly

Comment (shiv-am0) [reply plugin.go]: Removed the function and used it directly.

File: docker/Dockerfile.linux.arm64 (hunk 1 above)
Comment (shiv-am0) [reply docker/Dockerfile.linux.arm64]: Resolved. Will make this change in a different PR.

File: plugin.go [hunk 4]
```diff
@@ -78,6 +83,73 @@ func (p *Plugin) Exec(client *storage.Client) error {
...
+
+		log.Println("Downloading objects ...")
+
+		ctx := context.Background()
+		query := &storage.Query{Prefix: p.Config.Source}
+
+		// List the objects in the specified GCS bucket path
+		it := p.bucket.Objects(ctx, query)
+
+		g := errgroup.Group{}
```
[the remark this answers is not in the document]
Comment (shiv-am0) [reply plugin.go]: Removed the goroutine and executed it in a sequential manner.

File: plugin.go (hunk 2 above)
Comment (shiv-am0) [reply plugin.go]: Introduced a function to perform downloading

File: main.go [hunk 5]
```diff
@@ -54,6 +54,11 @@ func main() {
 			Usage:  "destination to copy files to, including bucket name",
 			EnvVar: "PLUGIN_TARGET",
 		},
+		cli.BoolFlag{
+			Name:   "download",
+			Usage:  "switch to download mode, which will fetch `target`'s files from GCS",
```
Comment (vistaarjuneja) [inline main.go:59]: Nit: target -> source?

File: plugin.go [hunk 6]
```diff
@@ -33,6 +34,9 @@ type (
 		// Destination to copy files to, including bucket name
 		Target string
 
+		// if true, plugin is set to download mode, which means `target` from the bucket will be downloaded
```
Comment (vistaarjuneja) [inline plugin.go:37]: Nit `target` should be `source`?

File: plugin.go [hunk 7]
```diff
@@ -78,6 +82,21 @@ func (p *Plugin) Exec(client *storage.Client) error {
...
 	p.bucket = client.Bucket(strings.Trim(bname, "/"))
 
+	// If in download mode, call the Download method
+	if p.Config.Download {
+		bname, remainingPath := extractBucketName(p.Config.Source)
+		p.Config.Source = remainingPath
+
+		p.bucket = client.Bucket(strings.Trim(bname, "/"))
+
+		log.Println("Downloading objects ...")
```
Comment (vistaarjuneja) [inline plugin.go:92]: Nit: also log the bucket and path in this log - like `Downloading objects from bucket: %s using path: %s`

**Comment by vistaarjuneja** [comment]
Looks good! Just some final minor comments

File: main.go (hunk 5 above)
Comment (shiv-am0) [reply main.go]: Changed

File: plugin.go (hunk 6 above)
Comment (shiv-am0) [reply plugin.go]: Changed

File: plugin.go (hunk 7 above)
Comment (shiv-am0) [reply plugin.go]: Added log statement

**Review by vistaarjuneja — APPROVED** [review]
LGTM!

**Pull request merged**